### PR TITLE
Upgrade to node@4, jest@0.5

### PR DESCRIPTION
**Pull request by zpao**
Still some failing tests to work through

### CODE REVIEW

**Comment by zpao** [comment]
Ok rebased and started using node v4. I fixed 1 issue that was screwing up the rethrowing of errors. Now there's just the tests that are doing some hacky things to test full page rendering…

**Comment by zpao** [comment]
@cpojer - perhaps you know of a better way in jest/jsdom to effectively get a new standalone document - effectively what we did in https://github.com/facebook/react/blob/master/src/test/getTestDocument.js but less hacky so it still works…

**Comment by cpojer** [comment]
I would recommend writing separate test files if you want a different env.

**Comment by zpao** [comment]
Yea, I guess we could do that. It's a bit obnoxious, though maybe we can just use `document` itself now (some of this hackery might have been to support the "run in the browser" case). Thanks for the advice.

**Comment by spicyj** [comment]
I don't really want to make ReactErrorUtils behave differently with jest…

**Comment by zpao** [comment]
Updated and removed the ErrorUtils commit thanks to @spicyj fixing it better. I also added the cache key stuff for @cpojer

File: scripts/jest/environment.js [hunk 1]
```diff
@@ -1,2 +0,0 @@
-/* eslint-disable */
-__DEV__ = true;
```
Comment (cpojer) [inline scripts/jest/environment.js]: why did you remove this?

File: scripts/jest/preprocessor.js [hunk 2]
```diff
@@ -58,4 +63,6 @@ module.exports = {
     }
     return src;
   },
+
+  getCacheKey: createCacheKeyFunction([__filename]),
```
Comment (cpojer) [inline scripts/jest/preprocessor.js]: this should also include babel's package.json and the path to the two babel plugins from above.

File: package.json [hunk 3]
```diff
@@ -54,6 +54,9 @@
     "test": "NODE_ENV=test jest"
   },
   "jest": {
+    "globals": {
+      "__DEV__": true
```
Comment (zpao) [inline package.json]: @cpojer Because I added it here.

It was unclear to me if using `var` in environment would be in the global scope but there was no question about using this config.

File: scripts/jest/preprocessor.js (hunk 2 above)
Comment (zpao) [inline scripts/jest/preprocessor.js]: 👍 I thought I might.

File: package.json (hunk 3 above)
Comment (cpojer) [inline package.json]: yeah I saw and removed my comment. Didn't know we had this config, this kinda sucks because the config is not very useful. The env setup script exposes both `global` and `window` to which you can attach globals.

Comment (zpao) [inline package.json]: `global === window`? I'm happy to do that and would prefer the explicitness. The initial reason I moved it here was that my babel dev transform was trying to transform this and the quickest fix was to just move the definition.

Comment (cpojer) [inline package.json]: yeah it should be the same.

**Comment by spicyj** [comment]
lgtm with the cache key changes.

**Comment by cpojer** [comment]
:tada:

**Comment by zpao** [comment]
Now with cache key changes... (ps @cpojer - using `require.resolve` might be good for our other places)

**Comment by cpojer** [comment]
Ok, 39 seconds on my MBP on a cached (2nd) test run. Not fast enough but I think it is reasonable given the amount of tests. Does anyone have stats on what it was like before?

**Comment by spicyj** [comment]
For a second run, "time npm test" claims 27.5s before, 17.4s now on my machine.

**Comment by cpojer** [comment]
I guess you have a better machine than I do ;) Seems like a reasonable runtime, maybe we can get it down to 10s soon-ish.
